core/tests/puppeteer-acceptance-tests/puppeteer-testing-utilities/user-factory.ts

```diff
@@ -0,0 +1,169 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Utility File for declaring and initializing users.
+ */
+
+import { SuperAdminFactory, ISuperAdmin } from '../user-utilities/super-admin-utils';
+import { BaseUserFactory, IBaseUser } from './puppeteer-utils';
+import { TranslationAdminFactory } from '../user-utilities/translation-admin-utils';
+import { LoggedInUserFactory, ILoggedInUser } from '../user-utilities/logged-in-users-utils';
+import { BlogAdminFactory, IBlogAdmin } from '../user-utilities/blog-admin-utils';
+import { QuestionAdminFactory } from '../user-utilities/question-admin-utils';
+import { BlogPostEditorFactory } from '../user-utilities/blog-post-editor-utils';
+
+/**
+ * Global user instances that are created and can be reused again.
+ */
+let superAdminInstance: ISuperAdmin & IBlogAdmin | null = null;
+let activeUsers: IBaseUser[] = [];
+
+/**
+ * Mapping of user roles to their respective function class.
+ */
+const USER_ROLE_MAPPING = {
+  'translation admin': TranslationAdminFactory,
+  'blog admin': BlogAdminFactory,
+  'blog post editor': BlogPostEditorFactory,
+  'question admin': QuestionAdminFactory
+};
+
+/**
+ * These types are used to create a union of all the roles and then
+ * create an intersection of all the roles. This is used to create a
+ * composition of the user and the role for type inference.
+ */
+type UnionToIntersection<U> =
+  (U extends IBaseUser ? (k: U) => void : never) extends
+    ((k: infer I) => void) ? I : never;
+
+type MultipleRoleIntersection<T extends (keyof typeof USER_ROLE_MAPPING)[]> =
+  UnionToIntersection<ReturnType<typeof USER_ROLE_MAPPING[T[number]]>>;
+
+/**
+ * This function creates a composition of the user and the role
+ * through object prototypes and returns the instance of that user.
+ */
+let composeUserWithRole = function<
+  TUser extends IBaseUser,
+  TRole extends IBaseUser
+>(
+    user: TUser,
+    role: TRole
+): TUser & TRole {
+  const userPrototype = Object.getPrototypeOf(user);
+  const rolePrototype = Object.getPrototypeOf(role);
+
+  Object.getOwnPropertyNames(rolePrototype).forEach((name: string) => {
+    Object.defineProperty(
+      userPrototype,
+      name,
+      Object.getOwnPropertyDescriptor(rolePrototype, name) ||
+        Object.create(null)
+    );
+  });
+
+  return user as TUser & TRole;
+};
+
+/**
+ * This function assigns roles to a user and returns the instance of that user.
+ */
+export let assignRolesToUser = async function<
+  TUser extends IBaseUser,
+  TRole extends keyof typeof USER_ROLE_MAPPING
```

Comment:
I am a bit confused about TUser and TRole. Why do we name the types like that, rather than just User and Role?